macro/derive/src/lib.rs

```diff
@@ -297,42 +318,44 @@ fn impl_from_variant(ast: &FromVariantInput) -> TokenStream {
 
     let generics = &ast.generics;
 
-    let froms = ast
+    let enum_data = ast
         .data
         .as_ref()
         .take_enum()
-        .expect("BUG: this should've been checked when parsing FromVariantInput")
-        .into_iter()
-        .filter_map(|variant| {
-            if !variant.fields.is_newtype() {
-                return None;
-            }
-            let span = variant.span();
-            let field = variant
-                .fields
-                .iter()
-                .next()
-                .expect("BUG: this should've been checked when parsing FromVariantInput");
-            let variant_type = &field.ty;
-
-            let try_into = if field.skip_try_from {
-                quote!()
-            } else {
-                try_into_variant(span, name, &variant.ident, variant_type, generics)
-            };
-            let from = if field.skip_from {
-                quote!()
-            } else if field.skip_container {
-                from_variant(span, name, &variant.ident, variant_type, generics, true)
-            } else {
-                from_variant(span, name, &variant.ident, variant_type, generics, false)
-            };
-
-            Some(quote!(
-                #try_into
-                #from
-            ))
-        });
+        .expect("BUG: this should've been checked when parsing FromVariantInput");
+    let variant_count = enum_data.len();
+    let froms = enum_data.into_iter().filter_map(|variant| {
+        if !variant.fields.is_newtype() {
+            return None;
+        }
+        let span = variant.span();
+        let field = variant
+            .fields
+            .iter()
+            .next()
+            .expect("BUG: this should've been checked when parsing FromVariantInput");
```

Comment:
Not your code, but there's an opportunity for improvement. 

The panic message should act as a comment explaining what invariant must be enforced. Consider saying the **what**.